dask/dataframe/optimize.py

```diff
@@ -48,87 +48,189 @@ def optimize(dsk, keys, **kwargs):
-                # ... where this value is __getitem__...
-                return dsk
-
-            block_columns = block.indices[1][0]
-            if isinstance(block_columns, str):
-                if block_columns in layers.keys():
-                    # Not a column selection if the getitem
-                    # key is a collection key
-                    column_projection = False
-                    break
-                block_columns = [block_columns]
-            elif np.issubdtype(type(block_columns), np.integer):
-                block_columns = [block_columns]
-            columns |= set(block_columns)
-            update_blocks[dep] = block
+
+        # Bail on the optimization if the  IO layer is in the
+        # requested keys, because we cannot change the name
+        # anymore. These keys are structured like
+        # [('getitem-<token>', 0), ...] so we check for the
+        # first item of the tuple.
+        # (See https://github.com/dask/dask/issues/5893)
+        if any(
+            layers[io_layer_name].name == x[0] for x in keys if isinstance(x, tuple)
+        ):
+            continue
+
+        # Inspect dependents of the current IO layer
+        deps = dsk.dependents[io_layer_name]
+
+        # This optimization currently supports two variations
+        # of `io_layer_name` dependents (`deps`):
+        #
+        #  - CASE A
+        #    - 1 Dependent: A column-based getitem layer
+        #    - This corresponds to the simple case that a
+        #      column selection directly follows the IO layer
+        #
+        #  - CASE B
+        #    - >1 Dependents: An arbitrary number of column-
+        #      based getitem layers, and a single row selection
+        #    - Usually corresponds to a filter operation that
+        #      may (or may not) precede a column selection
+        #    - This pattern is typical in Dask-SQL SELECT
+        #      queries that include a WHERE statement
+
+        # Bail if dependent layer type(s) do not agree with
+        # case A or case B
+
+        if not all(_is_selection(dsk.layers[k]) for k in deps) or {
+            _kind(dsk.layers[k]) for k in deps
+        } not in (
+            {"column-selection"},
+            {"column-selection", "row-selection"},
+        ):
+            continue
+
+        # Split the column- and row-selection layers.
+        # For case A, we will simply use information
+        # from col_select_layers to perform column
+        # projection in the root IO layer. For case B,
+        # these layers are not the final column selection
+        # (they are only part of a filtering operation).
+        row_select_layers = {k for k in deps if _kind(dsk.layers[k]) == "row-selection"}
+        col_select_layers = deps - row_select_layers
+
+        # Can only handle single column-selection
+        # layer if there is no row selection (case A)
+        if len(row_select_layers) == 0 and len(col_select_layers) > 1:
+            continue
```

Comment:
TODO: I suppose this constraint is unnecessary. We can just take the set union of the multiple column-selection dependents.